python/utils/OpaqueArguments.h

```diff
@@ -185,6 +189,87 @@ inline std::string mlirTypeToString(mlir::Type ty) {
+inline std::pair<std::size_t, std::vector<std::size_t>>
+getTargetLayout(func::FuncOp func, cudaq::cc::StructType structTy) {
+  auto mod = func->getParentOfType<ModuleOp>();
+  StringRef dataLayoutSpec = "";
+  if (auto attr = mod->getAttr(cudaq::opt::factory::targetDataLayoutAttrName))
+    dataLayoutSpec = cast<StringAttr>(attr);
+  auto dataLayout = llvm::DataLayout(dataLayoutSpec);
+  // Convert bufferTy to llvm.
+  llvm::LLVMContext context;
+  LLVMTypeConverter converter(func.getContext());
+  cudaq::opt::initializeTypeConversions(converter);
+  auto llvmDialectTy = converter.convertType(structTy);
+  LLVM::TypeToLLVMIRTranslator translator(context);
+  auto *llvmStructTy =
+      cast<llvm::StructType>(translator.translateType(llvmDialectTy));
+  auto *layout = dataLayout.getStructLayout(llvmStructTy);
+  auto strSize = layout->getSizeInBytes();
+  std::vector<std::size_t> fieldOffsets;
+  for (std::size_t i = 0, I = structTy.getMembers().size(); i != I; ++i)
+    fieldOffsets.emplace_back(layout->getElementOffset(i));
+  return {strSize, fieldOffsets};
+}
+
+/// @brief For the current struct member variable type, insert the
+/// value into the dynamically-constructed struct.
+inline void handleStructMemberVariable(void *data, std::size_t offset,
+                                       Type memberType, py::object value) {
+  auto appendValue = [](void *data, auto &&value, std::size_t offset) {
+    std::memcpy(((char *)data) + offset, &value,
+                sizeof(std::remove_cvref_t<decltype(value)>));
+  };
+  llvm::TypeSwitch<Type, void>(memberType)
+      .Case([&](IntegerType ty) {
+        if (ty.getIntOrFloatBitWidth() == 1) {
+          appendValue(data, value.cast<py::bool_>(), offset);
+          return;
+        }
+        appendValue(data, (std::size_t)value.cast<py::int_>(), offset);
+      })
+      .Case([&](mlir::Float64Type ty) {
+        appendValue(data, (double)value.cast<py::float_>(), offset);
+      })
+      .Case([&](cudaq::cc::StdvecType ty) {
+        auto appendVectorValue = []<typename T>(py::object value, void *data,
+                                                std::size_t offset, T) {
+          auto asList = value.cast<py::list>();
+          std::vector<double> *values = new std::vector<double>(asList.size());
+          for (std::size_t i = 0; auto &v : asList)
+            (*values)[i++] = v.cast<double>();
+
+          std::memcpy(((char *)data) + offset, values, 16);
+        };
+
+        TypeSwitch<Type, void>(ty.getElementType())
+            .Case([&](IntegerType type) {
+              if (type.getIntOrFloatBitWidth() == 1) {
+                appendVectorValue(value, data, offset, bool());
+                return;
+              }
+
+              appendVectorValue(value, data, offset, std::size_t());
+              return;
+            })
+            .Case([&](FloatType type) {
+              if (type.getIntOrFloatBitWidth() == 32) {
```

Comment:
`type == builder.getF32Type()` might be a bit safer here. I think a 32 bit floating-point in MLIR is always IEEE single-precision at the moment, but it is possible that a collection of 32 bits might be use some other encoding. This already happens for half-precision formats.